cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -121,6 +131,10 @@ public Watch createWatch(
 
             return watch;
           }
+        } else if (hasClusterChanged && request.getTypeUrl().equals(Resources.ENDPOINT_TYPE_URL)) {
```

Comment:
This and response ordering seems to me like it should be in `AdsDiscoveryRequestStreamObserver` as it's directly related to how CP in ADS mode behaves and communicates with Envoy. Right now we pass information from `AdsDiscoveryRequestStreamObserver` so `SimpleCache` could make a decision. Indicator of this is the new javadoc on `hasClusterChanged` stating that's only relevant for ADS. I understand that's a fix for some issue but IMO we are making quite a mess in the codebase and we should think about the design a little bit.

Reply:
I am not sure if moving it to `AdsDiscoveryRequestStreamObserver` is even possible. This additional flag allows us to respond only when version of EDS hasn't changed. If we move logic to `AdsDiscoveryRequestStreamObserver` we won't have information about version of snapshot in cache so we won't be able to check if there is new version.